packages/core/src/layout/SignInPage/SignInPage.tsx

```diff
@@ -87,34 +87,63 @@ export const SingleSignInPage = ({
   const authApi = useApi(provider.apiRef);
   const configApi = useApi(configApiRef);
 
-  const [retry, setRetry] = useState<{} | boolean | undefined>(auto);
+  const [autoShowPopup, setAutoShowPopup] = useState<boolean>(auto ?? false);
+  // Defaults to true so that an initial check for existing user session is made
+  const [retry, setRetry] = useState<{} | boolean | undefined>(true);
   const [error, setError] = useState<Error>();
 
+  // The SignIn component takes some time to decide whether the user is logged-in or not.
+  // showLoginPage is used to prevent a glitch-like experience where the sign-in page is
+  // displayed for a split second when the user is already logged-in.
+  const [showLoginPage, setShowLoginPage] = useState<boolean>(false);
+
   useEffect(() => {
     const login = async () => {
-      const identity = await authApi.getBackstageIdentity({
-        instantPopup: true,
-      });
-
-      const profile = await authApi.getProfile();
-      onResult({
-        userId: identity!.id,
-        profile: profile!,
-        getIdToken: () => {
-          return authApi.getBackstageIdentity().then(i => i!.idToken);
-        },
-        signOut: async () => {
-          await authApi.signOut();
-        },
-      });
+      try {
+        let identity;
+        // Do an initial check if any logged-in session exists
+        identity = await authApi.getBackstageIdentity({
+          optional: true,
+        });
+
+        // If no session exists, show the sign-in page
+        if (!identity && autoShowPopup) {
+          // Unless auto is set to true, this step should not happen.
+          // When user intentionally clicks the Sign In button, autoShowPopup is set to true
+          identity = await authApi.getBackstageIdentity({
+            instantPopup: true,
+          });
+        }
+
+        if (!identity) {
+          setShowLoginPage(true);
+          return;
+        }
+
+        const profile = await authApi.getProfile();
+        onResult({
+          userId: identity!.id,
+          profile: profile!,
+          getIdToken: () => {
+            return authApi.getBackstageIdentity().then(i => i!.idToken);
+          },
+          signOut: async () => {
+            await authApi.signOut();
+          },
+        });
+      } catch (err) {
+        // User closed the sign-in modal
+        setError(err);
+        setShowLoginPage(true);
+      }
     };
 
     if (retry) {
```

Comment:
Retry is never false now, so you can get rid of this condition and initialize it to anything, like just undefined

Reply:
Yep!